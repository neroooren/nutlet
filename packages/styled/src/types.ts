import { Properties } from 'csstype'
import { DECLARATION, MEDIA, RULESET } from 'stylis'

export type Exact<T> = {
  [K in keyof T]: T[K]
}

// eslint-disable-next-line @typescript-eslint/ban-types
export type RecursivePartial<T> = T extends {} ? { [P in keyof T]?: RecursivePartial<T[P]> | undefined } : T
export type Expect<T, P> = T extends null | undefined ? never : P

export type CSSProps =
  | Properties<string | number>
  | {
      [key: string]: CSSProps
    }

export type VariantsDef<V> = {
  [name in keyof V]: { [variant in keyof V[name]]: CSSProps }
}

export type MorphVariant<T> = T extends number
  ? `${T}` | T
  : T extends 'true' | 'false'
  ? boolean
  : T extends `${number}`
  ? number | T
  : T

export type VariantProps<Variants> = {
  [name in keyof Variants]?: MorphVariant<keyof Variants[name]>
}

export type StyledProps<Variants, P = unknown> = Omit<P, keyof VariantProps<Variants> | 'as'> &
  VariantProps<Variants> & {
    as?: keyof JSX.IntrinsicElements
    className?: string
  }

export type StylesInit<Variants> = CSSProps & {
  variants?: VariantsDef<Variants>
}

export type Element = {
  line: number
  column: number
  length: number
  parent?: Element | null
  return: string
  root?: Element
  value: string
} & (
  | {
      type: typeof DECLARATION
      children: string
      props: string
    }
  | {
      type: typeof RULESET
      children: Element
      props: string[]
    }
  | {
      type: typeof MEDIA
      children: Element
      props: string[]
    }
)
