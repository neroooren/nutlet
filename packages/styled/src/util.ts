import hash from '@emotion/hash'

import { Expect } from './types'

const { create, assign, keys, entries } = Object

export { create, assign, keys, entries }

export const isBrowser = typeof window !== 'undefined' && typeof window.navigator !== 'undefined'

export function splitObject<P, G extends string>(
  obj: P,
  groupBy: (key: keyof P, val: P[typeof key]) => G | true | false,
): Expect<P, { [key in G]?: { [key in keyof P]?: P[key] } }> {
  if (!obj) {
    throw new Error('The object to be splitted must not be nullish.')
  }

  return entries(obj).reduce((groups, [key, val]) => {
    const group = groupBy(key as keyof P, val)
    if (groups[group]) {
      groups[group][key] = val
    } else {
      groups[group] = { [key]: val }
    }

    return groups
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  }, {} as any)
}

export function omit<P, T extends keyof NonNullable<P>>(obj: P, ...keys: T[]): Omit<P, T> {
  if (!obj) {
    return obj
  }

  // @ts-expect-error `splitObject` won't assure all possible banches are nullish
  return splitObject(obj, (key) => keys.includes(key as T)).false ?? {}
}

export function memoize<R>(func: (input: string) => R): { (input: string): R; cache: Map<string, R> } {
  const memoized = function memoizedFunc(this: unknown, input: string) {
    const cache = memoized.cache
    if (cache.has(input)) {
      return cache.get(input)!
    }

    const result = func.call(this, input)
    cache.set(input, result)
    return result
  }
  memoized.cache = new Map<string, R>()

  return memoized
}

export const hashString = memoize((input: string) => {
  return ('n' + hash(input)).substr(0, 6)
})

export function get(obj: unknown, path: string): unknown {
  if (!obj) {
    return undefined
  }

  if (!path) {
    return obj
  }

  return path.split('.').reduce((o, key) => {
    // @ts-expect-error object is unknown
    return o?.[key]
  }, obj)
}
