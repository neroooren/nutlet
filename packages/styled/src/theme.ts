import { createContext, createElement, FC, useContext } from 'react'

import { insertRules } from './sheet'
import { RecursivePartial } from './types'
import { entries, get, hashString } from './util'

// let user do declaration merge
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface Theme {}

type NutletTheme = Theme & {
  $$nutlet?: {
    className: string
    parent?: NutletTheme
  }
}

export const themeContext = createContext<Theme | undefined>(undefined)
// eslint-disable-next-line @typescript-eslint/naming-convention
export const ThemeComsumer = themeContext.Consumer
// eslint-disable-next-line @typescript-eslint/naming-convention
export const ThemeProvider: FC<{ theme: RecursivePartial<Theme> }> = ({ theme, children }) => {
  const parentTheme = useContext(themeContext)
  if (!('$$nutlet' in theme)) {
    const nutletTheme = initTheme(theme, parentTheme)

    const ruleset = getCSSVariablesRuleset(nutletTheme)
    insertRules([`.${nutletTheme.$$nutlet!.className}{${ruleset}}`])
  }

  return createElement(themeContext.Provider, { value: theme, children })
}

export function initTheme(theme: Theme, parent?: Theme): NutletTheme {
  if ('$$nutlet' in theme) {
    console.warn('Theme has already been initialized: ', theme)
    return theme
  }

  const hash = hashString(JSON.stringify(theme))

  ;(<NutletTheme>theme).$$nutlet = {
    // avoid hash starts with numbers which are not a valid selector in css
    className: `theme-${hash}`,
    parent: parent as NutletTheme,
  }

  return theme as NutletTheme
}

export function getCSSVariablesRuleset(theme: NutletTheme, scope = ''): string {
  let ruleset = ''
  if (!theme) {
    return ''
  }

  entries(theme).map(([key, val]) => {
    if (key === '$$nutlet') {
      return
    }
    if (typeof val === 'string') {
      if (!val.startsWith('$')) {
        ruleset += `--${scope}-${key}:${val};`
      }
    } else if (typeof val === 'number') {
      ruleset += `--${scope}-${key}:${val};`
    } else {
      ruleset += getCSSVariablesRuleset(val, scope ? `${scope}-${key}` : key)
    }
  })

  return ruleset
}

export function getThemeClassName(theme: NutletTheme): string {
  let $$nutlet = theme.$$nutlet
  const classNames: string[] = []

  while ($$nutlet) {
    classNames.unshift($$nutlet.className)
    $$nutlet = $$nutlet.parent?.$$nutlet
  }

  return classNames.join(' ')
}

const themeTokenPattern = /\$([a-zA-Z0-9.]+)/g
export function replaceThemeToken(theme: NutletTheme, value: string): string {
  let match: RegExpExecArray | null = null

  while ((match = themeTokenPattern.exec(value))) {
    const tokenPath = match[1]
    const val = get(theme, tokenPath)
    if (val) {
      let replacer: string
      if (typeof val === 'object') {
        // error
        break
      } else if (typeof val === 'string' && val.startsWith('$')) {
        // tokens use other token
        replacer = replaceThemeToken(theme, val)
      } else {
        replacer = `var(--${tokenPath.replaceAll('.', '-')})`
      }

      value = value.substring(0, match.index) + replacer + value.substring(match.index + tokenPath.length + 1)
    }
  }

  if (theme.$$nutlet?.parent && value.includes('$')) {
    return replaceThemeToken(theme.$$nutlet.parent, value)
  }

  return value
}
