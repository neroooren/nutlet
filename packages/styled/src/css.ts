import { compile, middleware, serialize, prefix, RULESET, DECLARATION, stringify } from 'stylis'

import { insertRules } from './sheet'
import { Theme, replaceThemeToken } from './theme'
import { StylesInit, CSSProps, Element } from './types'
import { keys, hashString, entries } from './util'

const unitlessKeys: { [key in keyof CSSProps]: 1 } = {
  animationIterationCount: 1,
  borderImageOutset: 1,
  borderImageSlice: 1,
  borderImageWidth: 1,
  boxFlex: 1,
  boxFlexGroup: 1,
  boxOrdinalGroup: 1,
  columnCount: 1,
  columns: 1,
  flex: 1,
  flexGrow: 1,
  msFlexPositive: 1,
  flexShrink: 1,
  gridRow: 1,
  gridRowEnd: 1,
  gridRowStart: 1,
  gridColumn: 1,
  gridColumnEnd: 1,
  gridColumnStart: 1,
  msGridRows: 1,
  msGridColumns: 1,
  fontWeight: 1,
  lineHeight: 1,
  opacity: 1,
  order: 1,
  orphans: 1,
  tabSize: 1,
  widows: 1,
  zIndex: 1,
  zoom: 1,
  // eslint-disable-next-line @typescript-eslint/naming-convention
  WebkitLineClamp: 1,

  // SVG-related properties
  fillOpacity: 1,
  floodOpacity: 1,
  stopOpacity: 1,
  strokeDasharray: 1,
  strokeDashoffset: 1,
  strokeMiterlimit: 1,
  strokeOpacity: 1,
  strokeWidth: 1,
}

export function css<Variants>(
  stylesInit: StylesInit<Variants>,
  theme?: Theme,
  variantProps: Record<string | number, unknown> = {},
): string {
  const classNames: string[] = []

  const { variants, ...cssInit } = stylesInit
  const blocks: Array<{ init: CSSProps; variant?: string; variantValue?: unknown }> = []

  if (keys(cssInit).length) {
    blocks.push({ init: cssInit })
  }

  if (variants) {
    Object.entries(variantProps).forEach(([key, value]) => {
      // @ts-expect-error type actually matches
      const variantInit = variants[key][value]
      if (variantInit) {
        blocks.push({ init: variantInit, variant: key, variantValue: value })
      }
    })
  }

  blocks.forEach((block) => {
    const stringifiedStyle = stringifyCSSProps(block.init)

    const hash = hashString(stringifiedStyle)
    const className = block.variant ? `${block.variant}-${block.variantValue}--${hash}` : hash

    insertRules(generateRules(className, stringifiedStyle, theme))
    classNames.push(className)
  })

  return classNames.join(' ')
}

function processNumericValue(key: string, val: string | number): string | number {
  if (!unitlessKeys[key as keyof CSSProps] && typeof val === 'number') {
    return `${val}px`
  }

  return val
}

export function stringifyCSSProps(props: CSSProps): string {
  return entries(props).reduce((str, [key, val]) => {
    if (val) {
      if (typeof val === 'object') {
        if (!Array.isArray(val)) {
          str += `${key}{${stringifyCSSProps(val)}}`
        }
      } else {
        str += `${key.replace(/[A-Z]|^ms/g, '-$&').toLowerCase()}:${processNumericValue(key, val)};`
      }
    }
    return str
  }, '')
}

const fixedElements = new WeakSet<Element>()
function compatSelector(element: Extract<Element, { type: typeof RULESET }>) {
  if (
    !element.parent ||
    // .length indicates if this rule contains pseudo or not
    !element.length
  ) {
    return
  }

  let parent = element.parent
  const value = element.value

  // if this is an implicitly inserted rule (the one eagerly inserted at the each new nested level)
  // then the props has already been manipulated beforehand as they that array is shared between it and its "rule parent"
  if (element.column === parent.column && element.line === parent.line) {
    return
  }

  while (parent.type !== RULESET) {
    if (parent.parent) {
      parent = parent.parent
    } else {
      return
    }
  }

  // short-circuit for the simplest case
  if (element.props.length === 1 && value.charCodeAt(0) !== /* colon */ 58 && !fixedElements.has(parent)) {
    return
  }

  fixedElements.add(element)

  let i = 0

  value.split(',').forEach((rawSelector) => {
    ;(parent.props as string[]).forEach((parentSelector) => {
      if (rawSelector.startsWith(':')) {
        element.props[i++] = parentSelector + rawSelector
      } else {
        element.props[i++] = rawSelector.includes('&\f')
          ? rawSelector.replace(/&\f/g, parentSelector)
          : `${parentSelector} ${rawSelector}`
      }
    })
  })
}

export function generateRules(className: string, initialStyle: string, theme?: Theme): string[] {
  const rules: string[] = []

  const normalizeMiddleware = (element: Element) => {
    if (element.type === RULESET) {
      compatSelector(element)
      // console.log(element)
    }

    if (element.type === DECLARATION) {
      let { children: value } = element

      if (theme) {
        value = replaceThemeToken(theme, value)
      }

      element.return = prefix(`${element.props}:${value};`, element.length)
      // console.log(element)
    }
  }

  serialize(
    compile(`.${className}{${initialStyle}`),
    middleware([
      // @ts-expect-error we use more accurate element type to reduce type casting and avoid useless runtime assertion
      normalizeMiddleware,
      stringify,
      (element) => {
        if (!element.root && element.return) {
          rules.push(element.return)
        }
      },
    ]),
  )

  return rules
}
