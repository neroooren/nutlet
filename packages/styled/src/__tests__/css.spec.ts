import { generateRules, stringifyCSSProps } from '../css'
import { initTheme } from '../theme'
import { CSSProps } from '../types'
import { entries } from '../util'

const theme = {
  colors: {
    blue: 'blue',
    yellow: 'yellow',
  },
}

const normal: CSSProps = {
  color: 'green',
  border: '1px solid gray',
  fontSize: 12,
}

const implicitParentSelector: CSSProps = {
  ':hover': {
    color: '$colors.blue',
  },
}

const complexImplicitParentSelector: CSSProps = {
  ...implicitParentSelector,
  ':hover': {
    '> :first-child': {
      fontWeight: 500,
    },
  },
}

const explicitParentSelector: CSSProps = {
  ...normal,
  '&:hover': { color: '$colors.blue' },
  ':active': { outline: '$colors.yellow' },
}

const parentSelectorCoExist: CSSProps = {
  ...normal,
  '&:hover': { '::before': { content: '":hover::before"', position: 'absolute' } },
  '::after': { '&:hover': { content: '"::after:hover"' } },
  '&, :hover': { '::before': { content: "'::before,:hover::before'" } },
  '& + &::after:': { '&:hover:': { content: '":hover+::after:hover"' } },
  '&, :not(:first-child)': { '::before, &::after': { content: '"complex"' } },
}

const prefix: CSSProps = {
  transform: 'rotate(180deg)',
  display: 'flex',
}

const cssProps = {
  normal,
  implicitParentSelector,
  complexImplicitParentSelector,
  explicitParentSelector,
  parentSelectorCoExist,
  prefix,
}

test('stringifyCSSProps', () => {
  expect(
    stringifyCSSProps({
      fontSize: 14,
      '*': {
        fontWeight: 400,
      },
    }),
  ).toBe('font-size:14px;*{font-weight:400;}')
})

test('generateRules', () => {
  initTheme(theme)

  entries(cssProps).forEach(([name, props]) => {
    expect(generateRules('n', stringifyCSSProps(props), theme)).toMatchSnapshot(`${name}: 
.n ${JSON.stringify(props, null, 2)}`)
  })
})
