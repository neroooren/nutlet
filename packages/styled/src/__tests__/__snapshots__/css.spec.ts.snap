// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateRules: complexImplicitParentSelector: 
.n {
  ":hover": {
    "> :first-child": {
      "fontWeight": 500
    }
  }
} 1`] = `
Array [
  ".n:hover >:first-child{font-weight:500;}",
]
`;

exports[`generateRules: explicitParentSelector: 
.n {
  "color": "green",
  "border": "1px solid gray",
  "fontSize": 12,
  "&:hover": {
    "color": "$colors.blue"
  },
  ":active": {
    "outline": "$colors.yellow"
  }
} 1`] = `
Array [
  ".n{color:green;border:1px solid gray;font-size:12px;}",
  ".n:hover{color:var(--colors-blue);}",
  ".n:active{outline:var(--colors-yellow);}",
]
`;

exports[`generateRules: implicitParentSelector: 
.n {
  ":hover": {
    "color": "$colors.blue"
  }
} 1`] = `
Array [
  ".n:hover{color:var(--colors-blue);}",
]
`;

exports[`generateRules: normal: 
.n {
  "color": "green",
  "border": "1px solid gray",
  "fontSize": 12
} 1`] = `
Array [
  ".n{color:green;border:1px solid gray;font-size:12px;}",
]
`;

exports[`generateRules: parentSelectorCoExist: 
.n {
  "color": "green",
  "border": "1px solid gray",
  "fontSize": 12,
  "&:hover": {
    "::before": {
      "content": "\\":hover::before\\"",
      "position": "absolute"
    }
  },
  "::after": {
    "&:hover": {
      "content": "\\"::after:hover\\""
    }
  },
  "&, :hover": {
    "::before": {
      "content": "'::before,:hover::before'"
    }
  },
  "& + &::after:": {
    "&:hover:": {
      "content": "\\":hover+::after:hover\\""
    }
  },
  "&, :not(:first-child)": {
    "::before, &::after": {
      "content": "\\"complex\\""
    }
  }
} 1`] = `
Array [
  ".n{color:green;border:1px solid gray;font-size:12px;}",
  ".n:hover::before{content:\\":hover::before\\";position:absolute;}",
  ".n::after:hover{content:\\"::after:hover\\";}",
  ".n::before,.n:hover::before{content:'::before,:hover::before';}",
  ".n+.n::after::hover:{content:\\":hover+::after:hover\\";}",
  ".n::before,.n:not(:first-child)::before,.n::after,.n:not(:first-child)::after{content:\\"complex\\";}",
]
`;

exports[`generateRules: prefix: 
.n {
  "transform": "rotate(180deg)",
  "display": "flex"
} 1`] = `
Array [
  ".n{-webkit-transform:rotate(180deg);-moz-transform:rotate(180deg);-ms-transform:rotate(180deg);transform:rotate(180deg);display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}",
]
`;
